File: cpp/examples/linux/comms/MCP2515/minimal/main.cpp
#include <cstdio>
#include <unistd.h>
#include "SPIConnectionLinux.h"
#include "MCP2515.h"

#ifndef TEST_SPI_BUS
#define TEST_SPI_BUS 0
#endif
#ifndef TEST_SPI_DEVICE
#define TEST_SPI_DEVICE 0
#endif

int main() {
    SPIConnectionLinux connection(TEST_SPI_BUS, TEST_SPI_DEVICE, 0, 10000000); // Create SPI connection, (bus=0, device=0, mode=0, max_speed_hz=10e6) → SPIConnectionLinux
    MCP2515Minimal mcp2515(connection);                                    // Construct and initialise the MCP2515, (connection, bitrate_kbps=125, osc_mhz=8) → MCP2515Minimal
                                                                            // initialises with default 125 kbit/s at 8 MHz, Normal mode

    uint8_t payload[4] = { 0xDE, 0xAD, 0xBE, 0xEF };
    uint8_t tx_buf = mcp2515.send(0x123, payload, 4);                      // Send a CAN frame, (id=0x123, data, len=4) → uint8_t buf_index
                                                                            // returns 0/1/2 (buffer used) or 0xFF (none free)

    CanFrame frame;
    bool got_frame = mcp2515.recv(frame, 100);                             // Poll for a received frame, (frame, timeout_ms=100) → bool
                                                                            // returns false on timeout, true on frame received

    printf("send_buf=%u recv=%d\n", tx_buf, got_frame ? 1 : 0);
    return 0;
}

File: rust/examples/linux/comms/mcp2515/minimal/src/main.rs
use embedded_hal_bus::spi::ExclusiveDevice;
use linux_embedded_hal::SpidevBus;
use periph::chips::comms::MCP2515Minimal;
use spidev::{SpiModeFlags, Spidev, SpidevOptions};

struct NullCs;
impl embedded_hal::digital::ErrorType for NullCs {
    type Error = core::convert::Infallible;
}
impl embedded_hal::digital::OutputPin for NullCs {
    fn set_low(&mut self) -> Result<(), Self::Error> { Ok(()) }
    fn set_high(&mut self) -> Result<(), Self::Error> { Ok(()) }
}

fn main() {
    let bus: u8 = std::env::var("SPI_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(0);
    let dev: u8 = std::env::var("SPI_DEVICE").ok().and_then(|v| v.parse().ok()).unwrap_or(0);
    let dev_path = format!("/dev/spidev{}.{}", bus, dev);
    let mut spi = Spidev::open(dev_path).expect("open spi");
    spi.configure(&SpidevOptions::new()
        .max_speed_hz(5_000_000)
        .mode(SpiModeFlags::SPI_MODE_0)
        .build()).expect("configure spi");
    let bus_obj = SpidevBus(spi);
    let device = ExclusiveDevice::new_no_delay(bus_obj, NullCs).expect("spi device");

    let mut chip = MCP2515Minimal::new(device, 125, 8).expect("init MCP2515");    // Create MCP2515 driver, (spi, bitrate_kbps=125, osc_mhz=8) → Result

    let buf = chip.send(0x123, &[0xDE, 0xAD, 0xBE, 0xEF], false).expect("send");  // Send standard frame, (id=0x123, data ≤8 B, extended=false) → Result<u8>
    println!("sent on TXB{}", buf);

    let _ = chip.recv(100).expect("recv");                                          // Poll for received frame, (timeout_ms=100) → Result<Option<CanFrame>>
    println!("recv ok");
}
